File: src/heap/heap.ts
/**
 *  MAX HEAP: T = O(log(n)), S = O(n) -- HEAP SORT: T = O(n log(n)), S = O(1)
 */

export class MaxHeap {
    heap: number[];
    heapSize: number;
    maxSize: number;

    constructor(h: number[] = [], max: number = 12){
        this.heap = this.heapify(h);
        this.heapSize = h.length;
        this.maxSize = h.length > max ? h.length : max;
    }

    insert(value: number): void {
        if(this.heapSize === this.maxSize)
            return console.log('full heap :(');

        this.heapSize++;
        this.heap[this.heapSize - 1] = value;
        this.siftUp(this.heapSize - 1);
    }

    siftUp(index: number): void {
        while(index > 0 && this.heap[this.getParent(index)] < this.heap[index]){
            this.swap(this.heap, index, this.getParent(index));
            index = this.getParent(index);
        }
    }

    siftDown(index: number, checkPt: number = Infinity): void {
        let heap = this.heap;
        let maxIndex = index;
        let left = this.getLeftChild(index);
        let right = this.getRightChild(index);

        if(left >= checkPt || right >= checkPt)
            return;

        if(left <= this.heapSize && heap[left] > heap[maxIndex])
            maxIndex = left;
        
        if(right <= this.heapSize && heap[right] > heap[maxIndex])
            maxIndex = right;

        if(index !== maxIndex){
            this.swap(heap, index, maxIndex);
            this.siftDown(maxIndex, checkPt);
        }
    }

    // - return max value in heap
    getMax(): number {
        if(!this.isEmpty())
            return this.heap[0];

        return -1;
    }

    // - returns true if heap contains no elements
    isEmpty(): boolean {
        return this.heapSize === 0;
    }

    // - returns the max item, removing it
    extractMax(): number {
        let result = this.heap[0];

        this.swap(this.heap, 0, (this.heapSize - 1));
        this.heapSize--;
        this.heap.length = this.heapSize; // remove the last leaf
        this.siftDown(0);

        return result;
    }

    // - removes item at index
    remove(index: number): void {
        this.heap[index] = Infinity;
        this.siftUp(index);
        this.extractMax();
    }

    // - create heap from given array items
    heapify(a: number[]): number[]{
        this.heap = a;
        this.heapSize = a.length;
        
        // - loop indexes of non-leaf nodes (half of the array are leaves)
        for(let i = ~~((this.heapSize / 2) - 1); i >= 0; i--)
            this.siftDown(i);

        return this.heap;
    }

    // - take an unsorted array and turn it into a sorted array in-place using a max heap -> O(n log n)
    heapSort(a: number[] = this.heap){
        let len = a.length - 1;

        if(this.isEmpty())
            this.heapify(a);
        
        for(let i = len; i >= 0; i--){
            this.swap(a, 0, i);
            this.siftDown(0, i); //todo: fix issue with i=0 swap
        }

    }

    private getParent(i: number): number {
        return ~~((i - 1) / 2);
    }

    private getLeftChild(i: number): number {
        return (2 * i) + 1;
    }
    
    private getRightChild(i: number): number {
        return (2 * i) + 2;
    }

    private swap(a: number[], x: number, y: number): void {
        let temp = a[x];
        a[x] = a[y];
        a[y] = temp;
    }

}

File: python/heap.py
# MAX HEAP: T = O(log(n)), S = O(n) -- HEAP SORT: T = O(n log(n)), S = O(1)
class MaxHeap:
    def __init__(self, h: list[int]) -> None:
        self.heap: list[int] = self.heapify(h)
        self.size = len(h) if len(h) > 0 else 0
        self.maxSize = len(h) if len(h) > 0 else 12

    def insert(self, v: int):
        if v > self.maxSize:
            return -1

        self.size += 1
        self.heap[len(self.heap)-1] = v
        self.siftUp(len(self.heap)-1)

    def siftUp(self, i: int):
        while i > 0 and self.heap[self.__getParent(i)] < self.heap[i]:
            self.heap[i], self.heap[self.__getParent(
                i)] = self.heap[self.__getParent(i)], self.heap[i]
            i = self.__getParent(i)

    def siftDown(self, i: int, ckpt=float('inf')):
        maxIndex = i
        left = self.__getLeftChild(i)
        right = self.__getRightChild(i)

        if left >= ckpt or right >= ckpt:
            return

        if left < self.size and self.heap[left] > self.heap[i]:
            maxIndex = left
        if right < self.size and self.heap[right] > self.heap[i]:
            if left < self.size and self.heap[right] > self.heap[left]:
                maxIndex = right

        if maxIndex != i:
            self.heap[maxIndex], self.heap[i] = self.heap[i], self.heap[maxIndex]
            self.siftDown(maxIndex, ckpt)

    def extractMax(self):
        res = self.heap[0]
        self.heap[0], self.heap[self.size -
                                1] = self.heap[self.size-1], self.heap[0]
        self.size -= 1
        self.heap.pop()
        self.siftDown(0)
        return res

    def remove(self, i: int):
        self.heap[i] = float('inf')
        self.siftUp(i)
        self.extractMax()

    def heapify(self, A: list[int]):
        self.heap = A
        self.size = len(A)

        for i in range(len(A)//2-1, -1, -1):
            self.siftDown(i)

        return self.heap

    def heapSort(self):
        if not self.heap:
            return

        for i in range(len(self.heap)-1, -1, -1):
            self.heap[0], self.heap[i] = self.heap[i], self.heap[0]
            self.siftDown(0, i)

    def __getParent(self, i: int):
        return (i-1)//2

    def __getRightChild(self, i: int):
        return (2*i)+2

    def __getLeftChild(self, i: int):
        return (2*i)+1


arr = [22, 2, 99, 1, 34, 13, 4, 5, 60, 77, 18]
myheap = MaxHeap(arr)
print(myheap.heap)
print(myheap.heapSort())
print(myheap.heap)
